egs/librispeech/asr/simple_v1/bpe_ctc_att_conformer_decode.py

```diff
@@ -6,83 +6,286 @@
 import argparse
 import logging
 import os
-import random
-import re
-import sys
 
+from collections import defaultdict
 from pathlib import Path
+from typing import Any
+from typing import Dict
+from typing import List
+from typing import Optional
 from typing import Union
 
 import k2
 import numpy as np
 import torch
 
-from snowfall.data import LibriSpeechAsrDataModule
+from k2 import Fsa, SymbolTable
+
 from snowfall.common import average_checkpoint, store_transcripts
+from snowfall.common import find_first_disambig_symbol
 from snowfall.common import get_texts
 from snowfall.common import load_checkpoint
 from snowfall.common import str2bool
 from snowfall.common import write_error_stats
+from snowfall.data import LibriSpeechAsrDataModule
+from snowfall.decoding.graph import compile_HLG
+from snowfall.decoding.lm_rescore import rescore_with_n_best_list
+from snowfall.decoding.lm_rescore import rescore_with_whole_lattice
+from snowfall.decoding.lm_rescore import compute_am_scores, _intersect_device
+from snowfall.models import AcousticModel
 from snowfall.models.conformer import Conformer
 from snowfall.text.numericalizer import Numericalizer
 from snowfall.training.ctc_graph import build_ctc_topo
+from snowfall.training.mmi_graph import get_phone_symbols
+
+
+def nbest_decoding(model, encoder_memory, memory_mask, lats: k2.Fsa, num_paths: int):
+    '''
+    N-best rescore with transformer-decoder model.
+    The basic idea is to first extra n-best paths from the given lattice.
+    Then extract word_seqs and token_seqs for each path.
+    Compute the negative log-likehood for each token_seq as 'language model score', called decoder_scores.
```

Comment:
Is it a typo here? Why is the log-likelihood **negative**?

Reply:
Actually **NOT** a typo. It's computed by torch.nn.functional.cross_entropy whose result is **negative** log-likelihood.

Reply:
According to the comment, decoder_scores is the negative log-likehood for each token_seq.
Can we remove `negative`?